crates/router/src/services/authentication.rs

```diff
@@ -1113,6 +1113,91 @@ where
     }
 }
 
+pub struct JWTAuthMerchantAndProfileFromRoute {
+    pub merchant_id: id_type::MerchantId,
+    pub profile_id: String,
+    pub required_permission: Permission,
+}
+
+impl JWTAuthMerchantAndProfileFromRoute {
+    pub fn new(
+        merchant_id_from_route: id_type::MerchantId,
+        permission: Permission,
+        profile_id_from_route: String,
+    ) -> Self {
+        Self {
+            merchant_id: merchant_id_from_route,
+            required_permission: permission,
+            profile_id: profile_id_from_route,
+        }
+    }
+}
+
+#[async_trait]
+impl<A> AuthenticateAndFetch<AuthenticationData, A> for JWTAuthMerchantAndProfileFromRoute
+where
+    A: SessionStateInfo + Sync,
+{
+    async fn authenticate_and_fetch(
+        &self,
+        request_headers: &HeaderMap,
+        state: &A,
+    ) -> RouterResult<(AuthenticationData, AuthenticationType)> {
+        let payload = parse_jwt_payload::<A, AuthToken>(request_headers, state).await?;
+        if payload.check_in_blacklist(state).await? {
+            return Err(errors::ApiErrorResponse::InvalidJwtToken.into());
+        }
+
+        if payload.merchant_id != self.merchant_id {
+            return Err(report!(errors::ApiErrorResponse::InvalidJwtToken));
+        }
+
+        crate::core::utils::validate_profile_id_from_auth_layer(
+            payload.profile_id.clone(),
+            &crate::core::utils::ProfileIdWrapper(self.profile_id.clone()),
+        )
+        .change_context(errors::ApiErrorResponse::InvalidJwtToken)?;
```

Comment:
Can we do normal if...else instead of this. We will make the profile_id non-optional in JWT later and this is not required if both of them are not optional.

Reply:
Currently the logic for validating profile_id from auth layer is a bit complex (involves matching on tuple of two options). I think it will be better to keep this logic within a function for now. 

Later when the profileId is made non-optional in JWT, we can simply change the logic within the utils function and not in other places.

Reply:
But the trait you created will have `profile_id` that is coming from some `Object` will still be `Option`. But this is not the case here.

After a point of time, this will just become
```
  if payload.profile_id != self.profile_id {
    return Err(errors::ApiErrorResponse::InvalidJwtToken);
  }
```

Which is very simple than the trait logic.